admin-dev/themes/new-theme/js/pages/product/edit/combinations-manager.ts

```diff
@@ -41,11 +42,43 @@ const CombinationEvents = ProductEventMap.combinations;
 const CombinationsMap = ProductMap.combinations;
 
 export default class CombinationsManager {
+  productId: number;
+
+  eventEmitter: typeof EventEmitter;
+
+  $productForm: JQuery;
+
+  $combinationsContainer: JQuery;
+
+  $externalCombinationTab: JQuery;
+
+  $preloader: JQuery;
+
+  $paginatedList: JQuery;
+
+  $emptyState: JQuery;
+
+  paginator!: DynamicPaginator;
```

Comment:
```suggestion
  paginator?: DynamicPaginator;
```

Reply:
It should be a `?` no because the field can be empty as it's not initialized by default And we need to detect when it's empty to create it
I thought the `!:` notation meant `it is defined don't worry!!` which isn't the case here

Reply:
Because the paginator is created right on time during the first init

Reply:
```suggestion
  paginator: DynamicPaginator | null;
```

also works but I believe it's equivalent to `?:` or is it rather?

```suggestion
  paginator: DynamicPaginator | undefined;
```

Reply:
Unless you consider that based on how the component is built when this field is called it will necessarily have been created and will always be present

In this case I agree the `!:` makes sense but then you don't need to have the `if this.paginator` conditions in the other methods